File: wixi.backend/wixi.Business/Concrete/ContentSettingsService.cs
using System;
using System.Threading.Tasks;
using Microsoft.Extensions.Caching.Memory;
using Microsoft.EntityFrameworkCore;
using wixi.Business.Abstract;
using wixi.DataAccess.Concrete.EntityFramework.Contexts;
using wixi.Entities.Concrete;

namespace wixi.Business.Concrete
{
    public sealed class ContentSettingsService : IContentSettingsService
    {
        private const string CacheKey = "content_settings_cache";
        private readonly WixiDbContext _db;
        private readonly IMemoryCache _cache;

        public ContentSettingsService(WixiDbContext db, IMemoryCache cache)
        {
            _db = db;
            _cache = cache;
        }

        public async Task<ContentSettings?> GetAsync()
        {
            if (_cache.TryGetValue(CacheKey, out ContentSettings cached))
                return cached;

            var settings = await _db.ContentSettings.AsNoTracking().FirstOrDefaultAsync();
            if (settings != null)
            {
                _cache.Set(CacheKey, settings, TimeSpan.FromMinutes(5));
            }
            return settings;
        }

        public async Task<ContentSettings> UpsertAsync(ContentSettings input, string? updatedBy = null)
        {
            var existing = await _db.ContentSettings.FirstOrDefaultAsync();
            if (existing == null)
            {
                input.UpdatedAt = DateTime.UtcNow;
                input.UpdatedBy = updatedBy;
                _db.ContentSettings.Add(input);
            }
            else
            {
                existing.FooterCompanyDescDe = input.FooterCompanyDescDe;
                existing.FooterCompanyDescTr = input.FooterCompanyDescTr;
                existing.FooterCompanyDescEn = input.FooterCompanyDescEn;
                existing.FooterCompanyDescAr = input.FooterCompanyDescAr;
                existing.FacebookUrl = input.FacebookUrl;
                existing.InstagramUrl = input.InstagramUrl;
                existing.TwitterUrl = input.TwitterUrl;
                existing.LinkedInUrl = input.LinkedInUrl;
                existing.AboutMissionText1De = input.AboutMissionText1De;
                existing.AboutMissionText1Tr = input.AboutMissionText1Tr;
                existing.AboutMissionText1En = input.AboutMissionText1En;
                existing.AboutMissionText1Ar = input.AboutMissionText1Ar;
                existing.AboutMissionText2De = input.AboutMissionText2De;
                existing.AboutMissionText2Tr = input.AboutMissionText2Tr;
                existing.AboutMissionText2En = input.AboutMissionText2En;
                existing.AboutMissionText2Ar = input.AboutMissionText2Ar;
                existing.ContactPhone = input.ContactPhone;
                existing.ContactEmail = input.ContactEmail;
                existing.AddressGermany = input.AddressGermany;
                existing.AddressTurkeyMersin = input.AddressTurkeyMersin;
                existing.AddressTurkeyIstanbul = input.AddressTurkeyIstanbul;
                existing.UpdatedAt = DateTime.UtcNow;
                existing.UpdatedBy = updatedBy;
            }

            await _db.SaveChangesAsync();
            InvalidateCache();
            return (await GetAsync())!;
        }

        public void InvalidateCache() => _cache.Remove(CacheKey);
    }
}

File: V4/src/ApiServices/services/ContentSettingsService.ts
import { BASE_URL } from '../config/api.config';
import { TokenService } from './TokenService';

export interface ContentSettings {
  id?: number;
  footerCompanyDescDe: string;
  footerCompanyDescTr: string;
  footerCompanyDescEn?: string;
  footerCompanyDescAr?: string;
  facebookUrl?: string;
  instagramUrl?: string;
  twitterUrl?: string;
  linkedInUrl?: string;
  aboutMissionText1De: string;
  aboutMissionText1Tr: string;
  aboutMissionText1En?: string;
  aboutMissionText1Ar?: string;
  aboutMissionText2De: string;
  aboutMissionText2Tr: string;
  aboutMissionText2En?: string;
  aboutMissionText2Ar?: string;
  contactPhone: string;
  contactEmail: string;
  addressGermany: string;
  addressTurkeyMersin: string;
  addressTurkeyIstanbul: string;
}

export interface PublicContentSettings {
  footerCompanyDesc: string;
  socialMedia: {
    facebook?: string;
    instagram?: string;
    twitter?: string;
    linkedin?: string;
  };
  aboutMission: {
    text1: string;
    text2: string;
  };
  contact: {
    phone: string;
    email: string;
    addresses: {
      germany: string;
      turkeyMersin: string;
      turkeyIstanbul: string;
    };
  };
}

class ContentSettingsService {
  private async getAuthHeaders(): Promise<HeadersInit> {
    const tokenService = TokenService.getInstance();
    const token = await tokenService.getToken();
    return {
      'Content-Type': 'application/json',
      ...(token ? { Authorization: `Bearer ${token}` } : {}),
    };
  }

  // Public endpoint (no auth required)
  async getPublicSettings(lang: string = 'de'): Promise<{ success: boolean; settings?: PublicContentSettings; error?: string }> {
    try {
      const response = await fetch(`${BASE_URL}/api/v1.0/content-settings?lang=${lang}`, {
        method: 'GET',
        headers: {
          'Content-Type': 'application/json',
        },
        credentials: 'include',
      });

      if (!response.ok) {
        if (response.status === 404) {
          return { success: false, error: 'Settings not found' };
        }
        throw new Error(`Failed to fetch content settings: ${response.statusText}`);
      }

      // Backend returns PublicContentSettings object directly
      const settings = await response.json() as PublicContentSettings;
      return { success: true, settings };
    } catch (error) {
      console.error('Error fetching content settings:', error);
      return { success: false, error: error instanceof Error ? error.message : 'Unknown error' };
    }
  }

  // Admin endpoint
  async getSettings(): Promise<{ success: boolean; settings?: ContentSettings; error?: string }> {
    try {
      const headers = await this.getAuthHeaders();
      const response = await fetch(`${BASE_URL}/api/v1.0/admin/content-settings`, {
        method: 'GET',
        headers,
        credentials: 'include',
      });

      if (!response.ok) {
        if (response.status === 404) {
          return { success: false, error: 'Settings not found' };
        }
        throw new Error(`Failed to fetch content settings: ${response.statusText}`);
      }

      const settings = await response.json();
      return { success: true, settings };
    } catch (error) {
      console.error('Error fetching content settings:', error);
      return { success: false, error: error instanceof Error ? error.message : 'Unknown error' };
    }
  }

  async updateSettings(data: Omit<ContentSettings, 'id' | 'updatedAt' | 'updatedBy'>): Promise<{ success: boolean; error?: string }> {
    try {
      const headers = await this.getAuthHeaders();
      const response = await fetch(`${BASE_URL}/api/v1.0/admin/content-settings`, {
        method: 'PUT',
        headers,
        credentials: 'include',
        body: JSON.stringify(data),
      });

      if (!response.ok) {
        throw new Error(`Failed to update content settings: ${response.statusText}`);
      }

      return { success: true };
    } catch (error) {
      console.error('Error updating content settings:', error);
      return { success: false, error: error instanceof Error ? error.message : 'Unknown error' };
    }
  }
}

export default new ContentSettingsService();

